File: andriod/src/CooperHybrid/src/com/codesharp/cooper/Repositories/ChangeLogRepository.java
package com.codesharp.cooper.Repositories;

import java.util.List;

import android.R.string;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;

import com.codesharp.cooper.ChangeLog;
import com.codesharp.cooper.ChangeLogDao;
import com.codesharp.cooper.Constant;
import com.codesharp.cooper.MainActivity;
import com.codesharp.cooper.TaskIdx;
import com.codesharp.cooper.TaskIdxDao;

import de.greenrobot.dao.WhereCondition;

public class ChangeLogRepository {
	
	private MainActivity _activity;
	private ChangeLogDao _changeLogDao;
	private SharedPreferences _sharedPrefs;
	
	public ChangeLogRepository(MainActivity activity) {
		this._activity = activity;
		this._changeLogDao = this._activity.getSession().getChangeLogDao();
		this._sharedPrefs = PreferenceManager.getDefaultSharedPreferences(this._activity.getApplicationContext());
	}
	
	public List<ChangeLog> getAllChangeLogByTemp()
	{
		List<ChangeLog> changeLogs = this._changeLogDao.queryBuilder()
				.where(ChangeLogDao.Properties.AccountId.eq(""))
				.list();
		return changeLogs;
	}

	public void updateTasklistIdByNewId(String oldId, String newId) {
		
		List<ChangeLog> changeLogs = this.getAllChangeLog(oldId);
		for (ChangeLog changeLog : changeLogs) {
			changeLog.setTasklistId(newId);
		}
		if(changeLogs.size() > 0)
			this._changeLogDao.updateInTx(changeLogs);
	}

	public void updateAllToSend(String tasklistId) {
		
		if(this._sharedPrefs.contains(Constant.USERNAME_KEY))
		{
			String username = this._sharedPrefs.getString(Constant.USERNAME_KEY, "");
			this._changeLogDao.queryBuilder()
					.where(ChangeLogDao.Properties.AccountId.eq(username)
							, ChangeLogDao.Properties.TasklistId.eq(tasklistId))
					.buildDelete()
					.executeDeleteWithoutDetachingEntities();
		}
		else
		{
			this._changeLogDao.queryBuilder()
					.where(ChangeLogDao.Properties.AccountId.eq("")
							, ChangeLogDao.Properties.TasklistId.eq(tasklistId))
					.buildDelete()
					.executeDeleteWithoutDetachingEntities();
		}
	}

	public void addChangeLog(String type, String taskId, String name, String value, String tasklistId) {
		
		ChangeLog changeLog = new ChangeLog();
		changeLog.setChangeType(type);
		changeLog.setDataid(taskId);
		changeLog.setName(name);
		changeLog.setValue(value);
		changeLog.setIsSend(false);
		changeLog.setTasklistId(tasklistId);
		if(this._sharedPrefs.contains(Constant.USERNAME_KEY))
		{
			String username = this._sharedPrefs.getString(Constant.USERNAME_KEY, "");
			changeLog.setAccountId(username);
		}
		else
		{
			changeLog.setAccountId("");
		}
		this._changeLogDao.insert(changeLog);
	}

	//批量更新Changelog记录
	public void updateChangeLogs(List<ChangeLog> changeLogs) {
		
		this._changeLogDao.updateInTx(changeLogs);
	}

	public List<ChangeLog> getAllChangeLog(String tasklistId) {
		
		List<ChangeLog> changeLogs = null;
		if(this._sharedPrefs.contains(Constant.USERNAME_KEY))
		{
			String username = this._sharedPrefs.getString(Constant.USERNAME_KEY, "");
			changeLogs = this._changeLogDao.queryBuilder()
					.where(ChangeLogDao.Properties.AccountId.eq(username)
							, ChangeLogDao.Properties.TasklistId.eq(tasklistId))
					.list();
		}
		else
		{
			changeLogs = this._changeLogDao.queryBuilder()
					.where(ChangeLogDao.Properties.AccountId.eq("")
							, ChangeLogDao.Properties.TasklistId.eq(tasklistId))
					.list();
		}
		return changeLogs;
	}
}

File: windows phone/src/Cooper.Repositories/ChangeLogRepository.cs
using System;
using System.Collections.Generic;
using System.IO.IsolatedStorage;
using System.Linq;
using System.Linq;
using System.Net;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Documents;
using System.Windows.Ink;
using System.Windows.Input;
using System.Windows.Media;
using System.Windows.Media.Animation;
using System.Windows.Shapes;
using Cooper.Core;
using Cooper.Core.Models;

namespace Cooper.Repositories
{
    public class ChangeLogRepository
    {
        private CooperContext _context;

        public ChangeLogRepository()
        {
            this._context = new CooperContext();
        }

        public List<ChangeLog> GetAllChangeLogByTemp()
        {
            List<ChangeLog> changeLogs = this._context.ChangeLogs
                .Where(o => o.AccountId.Equals(""))
                .ToList();
            return changeLogs;
        }

        public List<ChangeLog> GetAllChangeLog(string tasklistId)
        {
            List<ChangeLog> changeLogs = null;
            if (IsolatedStorageSettings.ApplicationSettings.Contains(Constant.USERNAME_KEY))
            {
                string username = IsolatedStorageSettings.ApplicationSettings[Constant.USERNAME_KEY] as string;
                changeLogs = this._context.ChangeLogs
                    .Where(o => o.AccountId.Equals(username)
                        && o.TasklistId.Equals(tasklistId))
                    .ToList();
            }
            else
            {
                changeLogs = this._context.ChangeLogs
                    .Where(o => o.AccountId.Equals("")
                        && o.TasklistId.Equals(tasklistId))
                    .ToList();
            }
            return changeLogs;
        }

        public void UpdateTasklistIdByNewId(string oldId, string newId)
        {
            List<ChangeLog> changeLogs = this.GetAllChangeLog(oldId);
            foreach (var changeLog in changeLogs)
            {
                changeLog.TasklistId = newId;
            }
            if (changeLogs.Count > 0)
            {
                this._context.SubmitChanges();
            }
        }

        public void UpdateAllToSend(string tasklistId)
        {
            List<ChangeLog> changeLogs = null;
            if (IsolatedStorageSettings.ApplicationSettings.Contains(Constant.USERNAME_KEY))
            {
                string username = IsolatedStorageSettings.ApplicationSettings[Constant.USERNAME_KEY] as string;
                changeLogs = this._context.ChangeLogs
                    .Where(o => o.AccountId.Equals(username)
                        && o.TasklistId.Equals(tasklistId))
                    .ToList();
            }
            else
            {
                changeLogs = this._context.ChangeLogs
                    .Where(o => o.AccountId.Equals("")
                        && o.TasklistId.Equals(tasklistId))
                    .ToList();
            }
            if (changeLogs.Count > 0)
            {
                this._context.ChangeLogs.DeleteAllOnSubmit<ChangeLog>(changeLogs);
                this._context.SubmitChanges();
            }
        }

        public void AddChangeLog(string type, string taskId, string name, string value, string tasklistId)
        {
            ChangeLog changeLog = new ChangeLog();
            changeLog.ChangeType = type;
            changeLog.DataId = taskId;
            changeLog.Name = name;
            changeLog.Value = value;
            changeLog.IsSend = false;
            changeLog.TasklistId = tasklistId;
            if (IsolatedStorageSettings.ApplicationSettings.Contains(Constant.USERNAME_KEY))
            {
                string username = IsolatedStorageSettings.ApplicationSettings[Constant.USERNAME_KEY] as string;
                changeLog.AccountId = username;
            }
            else
            {
                changeLog.AccountId = "";
            }
            this._context.ChangeLogs.InsertOnSubmit(changeLog);
            this._context.SubmitChanges();
        }

        public void UpdateChangeLogs(List<ChangeLog> changeLogs)
        {
            this._context.SubmitChanges();
        }
    }
}
